picard/tagger.py

```diff
@@ -1246,12 +1246,52 @@ def sighandler(self):
         self.signalnotifier.setEnabled(True)
 
 
-def version():
-    print("%s %s %s" % (PICARD_ORG_NAME, PICARD_APP_NAME, PICARD_FANCY_VERSION_STR))
+class PicardArgs:
 
+    def __init__(self, argparse_args):
+        self.__load_from_argparse(argparse_args)
 
-def longversion():
-    print(versions.as_string())
+    def __load_from_argparse(self, argparse_args):
+        self.config_file = argparse_args.config_file
+        self.debug = argparse_args.debug
+        self._exec = argparse_args.exec
+        self.no_player = argparse_args.no_player
+        self.no_restore = argparse_args.no_restore
+        self.no_plugins = argparse_args.no_plugins
+        self.no_crash_dialog = argparse_args.no_crash_dialog
+        self.stand_alone_instance = argparse_args.stand_alone_instance
+        self._file_or_url = argparse_args.FILE_OR_URL
+        self.long_version = versions.as_string()
+
+        self.version = ""
+        self.processable = []
+
+        self.version = self.__get_version(short=argparse_args.version, long=argparse_args.long_version)
+        self.__parse_loadable_items()
+
+    def __parse_loadable_items(self):
+        for x in self._file_or_url:
+            if not urlparse(x).netloc:
+                x = os.path.abspath(x)
+            self.processable.append(x)
+
+        if self._exec:
+            for e in self._exec:
+                if "HELP" in [x.upper().strip() for x in e]:
+                    print_help_for_commands()
+                    sys.exit(0)
+                args = e[1:] if len(e) > 1 else ['']
+                for arg in args:
+                    self.processable.append(f"command://{e[0]} {arg}")
+
+        del self._exec
+        del self._file_or_url
```

Comment:
Those `del` are likely useless, since they point to data from `argparse`.

Try this example which tries to simulate what we have here:

`pip install memory_profiler`

```python
from memory_profiler import profile


class A:
    def __init__(self):
        self.v = "a" * 1000000


class B:
    def __init__(self, obj):
        self._v = obj.v

    def clear_v(self):
        del self._v


@profile
def main():
    obj_a = A()
    obj_b = B(obj_a)
    obj_b.clear_v()
    del obj_b
    del obj_a


if __name__ == '__main__':
    main()
```

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    17     35.2 MiB     35.2 MiB           1   @profile
    18                                         def main():
    19     36.3 MiB      1.0 MiB           1       obj_a = A()
    20     36.3 MiB      0.0 MiB           1       obj_b = B(obj_a)
    21     36.3 MiB      0.0 MiB           1       obj_b.clear_v()
    22     36.3 MiB      0.0 MiB           1       del obj_b
    23     35.4 MiB     -0.9 MiB           1       del obj_a
```

It is important to understand that Python only copies data if needed, and it uses references to this data.
In the case here, when you write `self._exec = argparse_args.exec` you just create a reference to the data initially stored by `argparse_args.exec`, so calling `del self._exec` don't actually free the memory used by the data, at least until you also `del argparse_args.exec`.

If we'd go for memory usage optimization it is rather elsewhere I'd look first, mainly in all the code processing data from MB or cover art for (possibly) millions of tracks... The code here is executed only once.

Note that I'm not against memory usage optimization, but to be clear, Python isn't the language for that... and if you really want to do that, use a memory profiler.

Reply:
Really well explained and helpful, thanks